components/org.wso2.carbon.identity.captcha/src/main/java/org/wso2/carbon/identity/captcha/connector/recaptcha/SSOLoginReCaptchaConfig.java

```diff
@@ -347,6 +366,45 @@ private boolean isReCaptchaEnabledForSSOLogin(ServletRequest servletRequest) {
             return false;
         }
 
+        if (containsOTPAuthenticatorAttributes(servletRequest) && !isOTPAsFirstFactor(context)) {
+            return false;
+        }
+
         return true;
     }
+
+    private boolean containsBasicAuthenticatorAttributes(ServletRequest servletRequest) {
+
+        return servletRequest.getParameter(USERNAME) != null && servletRequest.getParameter(PASSWORD) != null;
+    }
+
+    private boolean containsOTPAuthenticatorAttributes(ServletRequest servletRequest) {
+
+        return servletRequest.getParameter(USERNAME) != null && servletRequest.getParameter(OTPCODE) != null;
+    }
+
+    private boolean isOTPAsFirstFactor(AuthenticationContext context) {
+
+        return (context.getCurrentStep() == 1 || isPreviousIdPAuthenticationFlowHandler(context));
+    }
+
+    /**
+     * This method checks if all the authentication steps up to now have been performed by authenticators that
+     * implements AuthenticationFlowHandler interface. If so, it returns true.
+     * AuthenticationFlowHandlers may not perform actual authentication though the authenticated user is set in the
+     * context. Hence, this method can be used to determine if the user has been authenticated by a previous step.
+     *
+     * @param context   AuthenticationContext.
+     * @return True if all the authentication steps up to now have been performed by AuthenticationFlowHandlers.
+     */
+    private boolean isPreviousIdPAuthenticationFlowHandler(AuthenticationContext context) {
+
+        Map<String, AuthenticatedIdPData> currentAuthenticatedIdPs = context.getCurrentAuthenticatedIdPs();
+        return currentAuthenticatedIdPs != null && !currentAuthenticatedIdPs.isEmpty() &&
```

Comment:
When `currentAuthenticatedIdPs ==null` also, this will return a `false`. In that case, can we use this method to verify the authenticated user?

Reply:
Yes, this method is only used to determine if a user has been exclusively authenticated by AuthenticationFlowHandler instances. If the `currentAuthenticatedIdPs` is null, then that condition is false.
However, recaptcha verification will also happen when the current step number is 1 (that is also when the `currentAuthenticatedIdPs`  should be null).